mio/src/lib.rs

```diff
@@ -1,111 +1,129 @@
 extern crate mio;
 extern crate futures;
+extern crate fnv;
 
+use std::hash::BuildHasherDefault;
 use std::collections::HashMap;
 use std::io::{self, ErrorKind, Read, Write};
 use std::net::SocketAddr;
 use std::panic;
 use std::slice;
 use std::sync::Arc;
+use std::sync::atomic::{AtomicUsize, Ordering};
 use std::sync::mpsc::{channel, TryRecvError};
 
-use futures::{Future, promise, Complete, PollError};
+use fnv::FnvHasher;
 
-pub type IoFuture<T> = Future<Item=T, Error=io::Error>;
+use futures::{Future, Tokens, promise, Complete, Wake, PollError, PollResult};
 
-pub struct Loop {
-    io: mio::Poll,
-    tx: mio::channel::Sender<Message>,
-    rx: mio::channel::Receiver<Message>,
-    next: usize,
-    done: HashMap<usize, Complete<(), io::Error>>,
-}
-
-enum Message {
-    Wait(Complete<(), io::Error>, mio::EventSet, Arc<mio::Evented + Send + Sync>),
-    Register(Arc<mio::Evented + Send + Sync>),
-}
+pub type IoFuture<T> = Future<Item=T, Error=io::Error>;
 
+#[derive(Clone)]
```

Comment:
In the long run we probably don't want `impl Clone for TcpListener` I think, it seems like it could get into some uncomfortably weird situations if you can silently clone an I/O object. (the actual impl details here look good to me though)